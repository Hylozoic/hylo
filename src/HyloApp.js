export const JOINED_GROUP = 'JOINED_GROUP'
export const LEFT_GROUP = 'LEFT_GROUP'
export const NAVIGATION = 'NAVIGATION'
export const EDITOR = {
  onChange: 'onChange',
  onAddTopic: 'onAddTopic',
  onEnter: 'onEnter',
  setContent: 'setContent'
}

export function sendMessageToWebView (type, data) {
  if (!(window && window.ReactNativeWebView)) {
    throw new Error('`window.ReactNativeWebView` was not found')
  }

  if (!type) {
    throw new Error('Must provide a message `type` when sending a message to the WebView')
  }

  window.ReactNativeWebView.postMessage(JSON.stringify({ type, data }))
}

export function sendMessageFromWebView (webViewRef, type, data) {
  if (!(webViewRef && webViewRef.current && webViewRef.current.postMessage)) {
    throw new Error('The first parameter `webViewRef` is empty or not valid')
  }

  if (!type) {
    throw new Error('Must provide a message `type` when sending a message from the WebView')
  }

  webViewRef.current.postMessage(JSON.stringify({ type, data }))
}

export function parseWebViewMessage (messageOrEvent) {
  return JSON.parse(messageOrEvent?.nativeEvent ? messageOrEvent.nativeEvent.data : messageOrEvent)
}
