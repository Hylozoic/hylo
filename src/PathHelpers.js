export function topicPath (topicName, groupSlug) {
  if (groupSlug && !['all', 'public'].includes(groupSlug)) {
    return `/groups/${groupSlug}/topics/${topicName}`
  } else {
    return `/all/topics/${topicName}`
  }
}

export function mentionPath (memberId, groupSlug) {
  if (groupSlug) {
    return `/groups/${groupSlug}/members/${memberId}`
  } else {
    return `/members/${memberId}`
  }
}
