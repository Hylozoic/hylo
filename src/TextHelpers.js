import { marked } from 'marked'
import insane from 'insane'
import truncHtml from 'trunc-html'
import prettyDate from 'pretty-date'
import moment from 'moment-timezone'
import linkify from './linkify'

export function sanitize (text, whitelist, attrWhitelist) {
  if (!text) return ''
  if (whitelist && !Array.isArray(whitelist)) return ''

  // remove leading &nbsp; (a side-effect of contenteditable)
  const strippedText = text.replace(/<p>&nbsp;/gi, '<p>')

  return insane(strippedText, {
    allowedTags: whitelist || ['a', 'br', 'em', 'li', 'ol', 'p', 'strong', 'ul' ],
    allowedAttributes: attrWhitelist || {
      a: ['href', 'data-user-id', 'data-entity-type', 'target']
    }
  })
}

export function present (text, opts = {}) {
  const { slug, noLinks, maxlength, noP } = opts

  if (!text) return ''
  // wrap in a <p> tag, do this by default, require opt out
  if (text.substring(0, 3) !== '<p>' && !noP) text = `<p>${text}</p>`
  // make links and hashtags
  if (!noLinks) text = linkify(text, slug)
  if (maxlength) text = truncate(text, maxlength)

  return text
}

export const truncate = (text, length) => {
  return truncHtml(text, length, {
    sanitizer: {
      allowedAttributes: {
        a: ['href', 'class', 'data-search']
      }
    }
  }).html
}

export const markdown = text => {
  return sanitize(
    marked.parse(text || '', { gfm: true, breaks: true })
  )
}

export function humanDate (date, short) {
  const isString = typeof date === 'string'
  const isValidDate = !isNaN(Number(date)) && Number(date) !== 0
  let ret = date && (isString || isValidDate)
    ? prettyDate.format(isString ? new Date(date) : date)
    : ''

  if (short) {
    ret = ret.replace(' ago', '')
  } else {
    // this workaround prevents a "React attempted to use reuse markup" error
    // which happens if the timestamp is less than 1 minute ago, because the
    // server renders "N seconds ago", but by the time React is loaded on the
    // client side, it's "N+1 seconds ago"
    const match = ret.match(/(\d+) seconds? ago/)
    if (match) {
      if (Number(match[1]) >= 50) return '1m ago'
      return 'just now'
    }
  }

  return ret.replace(/ minutes?/, 'm')
    .replace(/ hours?/, 'h')
    .replace(/ days?/, 'd')
    .replace(/ weeks?/, 'w')
    .replace(/ month(s?)/, ' mo$1')
}

export const formatDatePair = (startTime, endTime, returnAsObj) => {
  const start = moment.tz(startTime, moment.tz.guess())
  const end = moment.tz(endTime, moment.tz.guess())

  const now = moment()
  const isThisYear = start.year() === now.year() && end.year() === now.year()

  let to = ''
  let from = ''

  if (isThisYear) {
    from = endTime ? start.format('ddd, MMM D [at] h:mmA') : start.format('ddd, MMM D [at] h:mmA z')
  } else {
    from = endTime ? start.format('ddd, MMM D, YYYY [at] h:mmA') : start.format('ddd, MMM D, YYYY [at] h:mmA z')
  }

  if (endTime) {
    if (end.year() !== start.year()) {
      to = end.format('ddd, MMM D, YYYY [at] h:mmA z')
    } else if (end.month() !== start.month() ||
               end.day() !== start.day() ||
               end <= now) {
      to = end.format('ddd, MMM D [at] h:mmA z')
    } else {
      to = end.format('h:mmA z')
    }
    to = returnAsObj ? to : ' - ' + to
  }

  return returnAsObj ? { from, to } : from + to
}

export function isDateInTheFuture (date) {
  return moment(date).isAfter(moment())
}
