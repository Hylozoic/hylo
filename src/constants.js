export const AnalyticsEvents = {
  COMMENT_CREATED: 'Comment Created',
  GROUP_CREATED: 'Group Created',
  GROUP_INVITATION_ACCEPTED: 'Group Invitation Accepted',
  DIRECT_MESSAGE_SENT: 'Direct Message Sent',
  POST_CREATED: 'Post Created',
  POST_UPDATED: 'Post Updated',
  TOPIC_CREATED: 'Topic Created',
  VOTED_ON_POST: 'Voted on Post',
  BLOCK_USER: 'User Blocked',
  UNBLOCK_USER: 'User Un-Blocked'
}

export const WebViewMessageTypes = {
  JOINED_GROUP: 'JOINED_GROUP',
  LEFT_GROUP: 'LEFT_GROUP',
  NAVIGATION: 'NAVIGATION',
  EDITOR: {
    BLUR: 'BLUR',
    CLEAR_CONTENT: 'CLEAR_CONTENT',
    FOCUS: 'FOCUS',
    LOADED: 'LOADED',
    ON_ADD_LINK: 'ON_ADD_LINK',
    ON_ADD_TOPIC: 'ON_ADD_TOPIC',
    ON_CHANGE: 'ON_CHANGE',
    ON_ENTER: 'ON_ENTER',
    SET_PROPS: 'SET_PROPS'
  }
}

// https://regex101.com/r/0M6mbp/1
export const HYLO_URL_REGEX = /^(https?:\/?\/?)?(www\.|staging\.)?(hylo\.com|localhost)(:?\d{0,6})(.*)/gi

// When/if ReduxORM models become shared merge add these back into the Group model
export const ALL_GROUPS_CONTEXT_SLUG = 'all'
export const PUBLIC_CONTEXT_SLUG = 'public'
