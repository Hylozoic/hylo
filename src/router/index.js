import React from 'react'
import { StaticRouter } from 'react-router'
import { ConnectedRouter } from 'connected-react-router'
import { createBrowserHistory, createMemoryHistory } from 'history'
import RootRouter from 'routes/RootRouter'
import '../css/global/index.scss'

export const history = typeof window !== 'undefined'
  ? createBrowserHistory()
  : createMemoryHistory()

export function clientRouter () {
  require('client/rollbar') // set up handling of uncaught errors

  return (
    <ConnectedRouter history={history}>
      <RootRouter />
    </ConnectedRouter>
  )
}

// Current SSR setup is deprecated and to be removed. See https://github.com/Hylozoic/hylo-evo/issues/1069
export function serverRouter (req, context) {
  return (
    <StaticRouter location={req.url} context={context}>
      <RootRouter />
    </StaticRouter>
  )
}
