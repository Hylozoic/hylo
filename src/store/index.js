import { createStore } from 'redux'
import createMiddleware from './middleware'
import { createBrowserHistory } from 'history'
import createRootReducer, { createCombinedReducers } from './reducers'

// `window` check here is for SSR case (see `appMiddleware`)
// where a static memory history is passed
export const history = typeof window !== 'undefined' && createBrowserHistory()

export function getEmptyState (providedHistory = history) {
  const combinedReducers = createCombinedReducers(providedHistory)

  return combinedReducers({}, { type: '' })
}

export default function configureStore (providedHistory = history, req) {
  return createStore(
    createRootReducer(providedHistory),
    getEmptyState(providedHistory),
    createMiddleware(providedHistory, req)
  )
}
