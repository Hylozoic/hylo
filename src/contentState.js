import { createElement } from 'react'
import Immutable from 'immutable'
import { convertToHTML, convertFromHTML } from 'draft-convert'
import { convertToRaw, convertFromRaw } from 'draft-js'
import { MENTION_ENTITY_TYPE, TOPIC_ENTITY_TYPE } from './constants'
import { detect } from 'detect-browser'

function serverDOMBuilder (html) {
  const jsdom = require('jsdom')
  const { JSDOM } = jsdom

  const { document: jsdomDocument, HTMLElement, HTMLAnchorElement } = (new JSDOM('<!DOCTYPE html>')).window
  // HTMLElement and HTMLAnchorElement needed on global for convertFromHTML to work
  global.HTMLElement = HTMLElement
  global.HTMLAnchorElement = HTMLAnchorElement
  // This was also needed at some point but doesn't seem to be any longer
  // global.Node = { TEXT_NODE: 3 }

  const doc = jsdomDocument.implementation.createHTMLDocument('foo')
  doc.documentElement.innerHTML = html
  const body = doc.getElementsByTagName('body')[0]
  return body
}

// NOTE: Legacy mention links are in this format --
// <a href="/u/99" data-user-id="99">Hylo User</a>
//
export function mentionFromLink (createEntity, node) {
  const mention = Immutable.Map({
    id: node.getAttribute('data-user-id'),
    name: node.text,
    avatar: ''
  })
  return createEntity(
    MENTION_ENTITY_TYPE,
    'SEGMENTED', // reference from plugin config?
    { mention }
  )
}

// NOTE: Legacy Topics are in this format --
// <a>#topic</a>
//
export function topicFromLink (createEntity, node) {
  const topic = Immutable.Map({
    name: node.text.substring(1)
  })
  return createEntity(
    TOPIC_ENTITY_TYPE,
    'IMMUTABLE', // reference from plugin config?
    { mention: topic }
  )
}

function mentionUrl (memberId, slug) {
  if (slug) {
    return `/c/${slug}/m/${memberId}`
  } else {
    return `/m/${memberId}`
  }
}

function mentionToLink (originalText, mention, slug) {
  return createElement(
    'a',
    {
      'data-entity-type': MENTION_ENTITY_TYPE,
      'data-user-id': mention.get('id'),
      href: mentionUrl(mention.get('id'), slug),
      className: 'mention'
    },
    originalText
  )
}

function tagUrl (tagName, slug) {
  if (slug) {
    return `/c/${slug}/tag/${tagName}`
  } else {
    return `/tag/${tagName}`
  }
}

function topicToLink (originalText, topic, slug) {
  return createElement(
    'a',
    {
      'data-entity-type': TOPIC_ENTITY_TYPE,
      'data-search': topic.get('name'),
      href: tagUrl(topic.get('name'), slug),
      className: 'hashtag'
    },
    originalText
  )
}

function externalLink (originalText, href) {
  return createElement(
    'a',
    {
      href,
      target: '_blank'
    },
    originalText
  )
}

export function fromHTML (html, { raw = true }) {
  const convertor = convertFromHTML(
    {
      htmlToEntity: (nodeName, node, createEntity) => {
        if (nodeName === 'a' && (node.getAttribute('data-entity-type') === MENTION_ENTITY_TYPE || node.getAttribute('data-user-id'))) {
          return mentionFromLink(createEntity, node)
        } else if (nodeName === 'a' && (node.getAttribute('data-entity-type') === TOPIC_ENTITY_TYPE || node.text[0] === '#')) {
          return topicFromLink(createEntity, node)
        }
      }
    }
  )
  // Use JSOM if in node, otherwise if in browser don't
  const { type } = detect()
  const contentState = convertor(html, {}, type === 'node' ? serverDOMBuilder : undefined)

  return raw
    ? convertToRaw(contentState)
    : contentState
}

export const toHTML = (unknownContentState, { slug }) => {
  // Automtically converts from raw format if not a contentState object
  const contentState = unknownContentState?.getEntityMap
    ? unknownContentState
    : convertFromRaw(unknownContentState)
  const result = convertToHTML({
    entityToHTML: (entity, originalText) => {
      switch (entity.type) {
        case MENTION_ENTITY_TYPE:
          return mentionToLink(originalText, Immutable.fromJS(entity.data.mention), slug)
        case TOPIC_ENTITY_TYPE:
          return topicToLink(originalText, Immutable.fromJS(entity.data.mention), slug)
        default:
          return originalText
      }
    }
  })(contentState)

  return result
}

export const toRaw = (unknownContentState) => {
  return unknownContentState?.getEntityMap
    ? convertToRaw(unknownContentState)
    : unknownContentState
}

export const fromRaw = (unknownContentState) => {
  return unknownContentState?.getEntityMap
    ? unknownContentState
    : convertFromRaw(unknownContentState)
}

// export function fromString (contentStateString) {
//   const contentStateRaw = JSON.parse(contentStateString)
//   return convertFromRaw(contentStateRaw)
// }
