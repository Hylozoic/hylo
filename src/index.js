import * as constants from './constants'
import * as text from './text'
import * as validators from './validators'
import * as contentState from './contentState'

export default {
  constants,
  text,
  validators,
  contentState
}
