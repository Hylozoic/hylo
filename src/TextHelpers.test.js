import * as TextHelpers from '../src/TextHelpers'
import moment from 'moment-timezone'


describe('processHTML', () => {
  it('converts Hylo.com URLs to relative hrefs', () => {
    const processResult = TextHelpers.processHTML(
      '<a href="https://www.hylo.com/groups/exit-to-community" target="_blank">https://www.hylo.com/groups/exit-to-community</a>',
    )
    expect(processResult).toBe(
      '<a href="/groups/exit-to-community" target="_self">https://www.hylo.com/groups/exit-to-community</a>'
    )
  })
})

describe('presentHTML', () => {
  it('truncates', () => {
    const presentResult = TextHelpers.presentHTML(
      '<p>Sadsadf <a href="#" data-entity-type="mention" data-user-id="26189">Matt Zeltzer Test</a> #test-topic  </p>\n',
      {
        slug: 'test',
        truncate: 20
      }
    )
    expect(presentResult).toBe(
      '<p>Sadsadf <a href="/groups/test/members/26189" data-entity-type="mention" data-user-id="26189" target="_self" class="mention">Matt …</a></p>'
    )
    const presentResultWithExtrenalLink = TextHelpers.presentHTML(
      '<p>Sadsadf <a href="https://google.com" target="_blank">GOOGLE</a></p>',
      {
        slug: 'test',
        truncate: 100
      }
    )
    expect(presentResultWithExtrenalLink).toBe(
      '<p>Sadsadf <a href="https://google.com" target="_blank">GOOGLE</a></p>'
    )
  })
})

describe('presentHTMLToText', () => {
  it("shouldn't include text of href on links", () => {
    expect(TextHelpers.presentHTMLToText("<a href='/any/url'>Text</a> more text")).toBe('Text more text')
  })
})

describe('truncateHTML', () => {
  it('has an ellipses after truncation', () => {
    expect(TextHelpers.truncateHTML('<a href="">test</a> test test', 5)).toBe('<a href="">test</a> …')
  })

  it('does not have an ellipses if there was no truncation', () => {
    expect(TextHelpers.truncateHTML('<a href="">test</a> test test', 100)).toBe('<a href="">test</a> test test')
  })
})

describe('truncateText', () => {
  it('has an ellipses after truncation', () => {
    expect(TextHelpers.truncateText('<I mean it> test test test', 8)).toBe('<I mean …')
  })

  it('does not have an ellipses if there was no truncation', () => {
    expect(TextHelpers.truncateText('<I mean it> test test test', 100)).toBe('<I mean it> test test test')
  })
})

describe('textLengthHTML', () => {
  it('should return lenght of plain text version of the html', () => {
    expect(TextHelpers.textLengthHTML('<strong>test</strong> <a href="">a link</a>')).toBe(11)
  })
})

describe('sanitizeHTML', () => {
  it('returns empty string if called without text', () => {
    expect(TextHelpers.sanitizeHTML()).toBe('')
  })

  it('allows whitelist to be undefined', () => {
    expect(TextHelpers.sanitizeHTML('foo')).toBe('foo')
  })

  it('strips leading whitespace in paragraphs', () => {
    expect(TextHelpers.sanitizeHTML('<p>&nbsp;</p>')).toBe('<p></p>')
  })

  it('removes tags not on a whitelist', () => {
    const expected = 'Wombats are great.<div>They poop square.</div>'
    const unsafe = 'Wombats are great.<em>So great.</em><div>They poop square.</div>'
    const actual = TextHelpers.sanitizeHTML(unsafe, { allowedTags: ['div'] })
    expect(actual).toBe(expected)
  })

  it('removes attributes not on a whitelist', () => {
    const expected = '<p id="wombat-data">Wombats are great.</p>'
    const unsafe = '<p id="wombat-data" class="main-wombat">Wombats are great.</p>'
    const actual = TextHelpers.sanitizeHTML(unsafe, { allowTags: ['p'], allowedAttributes: { p: ['id'] } })
    expect(actual).toBe(expected)
  })
})

describe('markdown', () => {
  it('converts to markdown', () => {
    expect(TextHelpers.markdown('*strong* **italic**')).toBe('<p><em>strong</em> <strong>italic</strong></p>\n')
  })
  it('converts to markdown not autolinking', () => {
    expect(TextHelpers.markdown('https://www.hylo.com')).toBe('<p>https://www.hylo.com</p>\n')
  })
  it('converts to markdown in paragraphs', () => {
    expect(TextHelpers.markdown('asdw\n\n\nasdf')).toBe('<p>asdw</p>\n<p>asdf</p>\n')
  })
})

describe('formatDatePair', () => {
  it('displays differences of dates', () => {
    const d1 = moment.tz(1551908483315, 'Etc/GMT').month(1).day(1).hour(18)
    const d2 = moment.tz(d1, 'Etc/GMT').hour(21)
    const d3 = moment.tz(d2, 'Etc/GMT').day(2)
    const d4 = moment.tz(d3, 'Etc/GMT').month(2)
    const d5 = moment.tz(d3, 'Etc/GMT').year(2050)

    expect(TextHelpers.formatDatePair(d1)).toMatchSnapshot()
    expect(TextHelpers.formatDatePair(d1, d2)).toMatchSnapshot()
    expect(TextHelpers.formatDatePair(d1, d3)).toMatchSnapshot()
    expect(TextHelpers.formatDatePair(d1, d4)).toMatchSnapshot()
    expect(TextHelpers.formatDatePair(d1, d5)).toMatchSnapshot()
  })
})
