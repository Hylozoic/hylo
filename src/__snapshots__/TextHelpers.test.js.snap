// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatDatePair displays differences of dates 1`] = `"Mon, Feb 4, 2019 at 11:41AM MST"`;

exports[`formatDatePair displays differences of dates 2`] = `"Mon, Feb 4, 2019 at 11:41AM - Mon, Feb 4 at 2:41PM MST"`;

exports[`formatDatePair displays differences of dates 3`] = `"Mon, Feb 4, 2019 at 11:41AM - Tue, Feb 5 at 2:41PM MST"`;

exports[`formatDatePair displays differences of dates 4`] = `"Mon, Feb 4, 2019 at 11:41AM - Tue, Mar 5 at 2:41PM MST"`;

exports[`formatDatePair displays differences of dates 5`] = `"Mon, Feb 4, 2019 at 11:41AM - Sat, Feb 5, 2050 at 2:41PM MST"`;
